### PR TITLE
Issue 245: Async token refresh

### LINKED ISSUE #245: Keycloak authentication token does not get refreshed
**Problem description**
An application that uses the byte stream reader to read from Pravega with Keycloak authentication works well for 5 minutes and then it panics. This is likely caused by not refreshing the token.

```
thread 'tokio-runtime-worker' panicked at 'get endpoint for async segment reader: RetryError { error: OperationError { can_retry: false, operation: "get_endpoint", error_msg: "status: Unauthenticated, message: \"\", details: [], metadata: MetadataMap { headers: {\"content-type\": \"application/grpc\", \"authorization\": \"Bearer snipped\"} }" }, total_delay: 0ns, tries: 1 }', /mnt/home/faheyc/nautilus/gstreamer/gstreamer-pravega/pravega-client-rust/src/segment_reader.rs:229:30
stack backtrace:
   0: rust_begin_unwind
             at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/std/src/panicking.rs:493:5
   1: core::panicking::panic_fmt
             at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/core/src/panicking.rs:92:14
   2: core::option::expect_none_failed
             at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/core/src/option.rs:1300:5
   3: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
   4: <tokio::future::poll_fn::PollFn<F> as core::future::future::Future>::poll
   5: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
   6: std::panicking::try
   7: tokio::runtime::task::harness::poll_future
   8: tokio::runtime::task::harness::Harness<T,S>::poll
   9: std::thread::local::LocalKey<T>::with
  10: tokio::runtime::thread_pool::worker::Context::run_task
  11: tokio::runtime::thread_pool::worker::Context::run
  12: tokio::macros::scoped_tls::ScopedKey<T>::set
  13: tokio::runtime::thread_pool::worker::run
  14: tokio::loom::std::unsafe_cell::UnsafeCell<T>::with_mut
  15: <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
  16: tokio::runtime::task::harness::Harness<T,S>::poll
  17: tokio::runtime::blocking::pool::Inner::run

```

**Problem location**
Keycloak client.

**Suggestions for an improvement**
None.

### PR DESCRIPTION
**Pull request by Tristan1900**
**Change log description**  
Currently token refresh is using a Runtime which makes it impossible to work in a async context. Use async way to refresh the token instead.

**Purpose of the change**  
Fix #245 

**What the code does**  
Use async way to refresh token

### CODE REVIEW

File: config/src/credentials.rs [hunk 1]
```diff
@@ -124,17 +123,16 @@ struct KeyCloak {
 
 #[async_trait]
 impl Cred for KeyCloak {
-    fn get_request_metadata(&self) -> String {
+    async fn get_request_metadata(&self) -> String {
         if self.is_expired() {
-            self.refresh_rpt_token();
+            self.refresh_rpt_token().await;
```
Comment (shrids) [inline config/src/credentials.rs:128]: A token refresh will be initiated only when there are about 5 seconds for the token to expire. But this method is not invoked automatically every time a request is made. Shouldn't it be the case?

File: controller-client/src/lib.rs [hunk 2]
```diff
@@ -556,12 +556,28 @@ impl ControllerClientImpl {
     /// which runs the connection in a background task and provides a `mpsc` channel interface.
     /// Due to this cloning the `Channel` type is cheap and encouraged.
     ///
-    fn get_controller_client(&self) -> ControllerServiceClient<Channel> {
-        self.channel.read().unwrap().clone()
+    async fn get_controller_client(&self) -> ControllerServiceClient<Channel> {
+        if self.config.is_auth_enabled && self.config.credentials.is_expired() {
+            // get_request_metadata internally checks if token expired before sending request to the server,
+            // race condition might happen here but eventually only one request will be sent.
+            let mut x = self.channel.write().await;
```
Comment (shrids) [inline controller-client/src/lib.rs:563]: we should remove code duplication of this logic here and `reset()`.

**Review by shrids — APPROVED** [review]
Apart from the code duplication comment the changes look good.

File: controller-client/src/lib.rs (hunk 2 above)
Comment (Tristan1900) [reply controller-client/src/lib.rs:545]: Thanks!

Comment (Tristan1900) [reply controller-client/src/lib.rs:563]: Just realize that we cannot remove the logic in reset. The reason is that `basic` method will never expire so if we remove the logic in reset then after resetting the auth header will be gone.

File: wire_protocol/src/connection.rs [hunk 3]
```diff
@@ -175,6 +175,15 @@ impl Connection for TlsConnection {
...
             .write_all(payload)
             .await
+            .context(SendData {
+                endpoint: endpoint.clone(),
+            })?;
+
+        self.stream
+            .as_mut()
+            .expect("get connection")
+            .flush()
```
Comment (shrids) [inline wire_protocol/src/connection.rs:185]: This API Flushes this stream, ensuring that all intermediately buffered contents reach their destination.
https://docs.rs/tokio-rustls/0.22.0/src/tokio_rustls/client.rs.html#93-94 snippet indicates a `TlsStream.flush()` is required to ensure data actually reaches the destination. 
Is this different with a TcpStream ?

Comment (Tristan1900) [reply wire_protocol/src/connection.rs:185]: yeah I think TcpStream itself doesn't have a buffer and any write is directly sent over the wire. However this TlsStream has a buffer inside that needs to be flushed. The name `TlsStream` is kind of misleading.

**Pull request merged**